### PR TITLE
[Merged by Bors] - feat(Dynamics/Birkhoff): define Birkhoff sum and Birkhoff average

### PR DESCRIPTION
**Pull request by urkud**
---

Should I add more properties to this PR?

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Dynamics/BirkhoffSum/Average.lean [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    birkhoffAverage R f g n x = birkhoffAverage S f g n x :=
+  map_birkhoffAverage R S (AddMonoidHom.id M) f g n x
+
+theorem birkhoffAverage_congr_ring' (S : Type _) [DivisionSemiring S] [Module S M] :
+    birkhoffAverage (α := α) (M := M) R = birkhoffAverage S := by
+  ext; apply birkhoffAverage_congr_ring
+
+end birkhoffAverage
+
+/-- Birkhofff average is "almost invariant" under `f`:
```
Comment (sgouezel) [inline Mathlib/Dynamics/BirkhoffSum/Average.lean:72]:
```suggestion
/-- Birkhoff average is "almost invariant" under `f`:
```

File: Mathlib/Dynamics/BirkhoffSum/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2023 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Algebra.BigOperators.Basic
+import Mathlib.Dynamics.FixedPoints.Basic
+
+/-!
+# Birkhoff sum and Birkhoff average
```
Comment (sgouezel) [inline Mathlib/Dynamics/BirkhoffSum/Basic.lean:10]:
```suggestion
# Birkhoff sums
```

File: Mathlib/Dynamics/BirkhoffSum/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+In this file we define `birkhoffSum f g n x` to be the sum `∑ k in Finset.range n, g (f^[k] x)`.
+This sum (more precisely, the corresponding average `n⁻¹ • birkhoffSum f g n x`)
+appears in various ergodic theorems
+saying that these averages converge to the "space average" `⨍ x, g x ∂μ` in some sense.
+
+See also `birkhoffAverage` defined elsewhere.
```
Comment (sgouezel) [inline Mathlib/Dynamics/BirkhoffSum/Basic.lean:17]: Can you name the file where it is defined?

**Review by sgouezel — COMMENTED** [review]
LGTM, thanks!
bors d+

File: Mathlib/Dynamics/BirkhoffSum/Basic.lean (hunk 3 above)
Comment (urkud) [inline Mathlib/Dynamics/BirkhoffSum/Basic.lean:17]:
```suggestion
See also `birkhoffAverage` defined in `Dynamics/BirkhoffSum/Average`.
```

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**